packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
+            return this.cachedDirectories.get(parentDirectory);
+        } else if (!this.pendingResolvedDirectories.has(parentDirectory)) {
+            const pendingDirectories = this.createResolutionPromise(parentDirectory);
+            this.pendingResolvedDirectories.set(parentDirectory, { input, pendingDirectories: pendingDirectories });
+        } else if (this.pendingResolvedDirectories.has(parentDirectory)) {
+            // if promise has still not resolved, update its input value to latest
+            const pendingDirectoriesData = (this.pendingResolvedDirectories.get(parentDirectory));
+            if (pendingDirectoriesData) {
+                pendingDirectoriesData.input = input;
+            }
+        }
+        return undefined;
+    }
+
+    protected async createResolutionPromise(directoryToResolve: string): Promise<void> {
+        return this.fileService.resolve(new URI(directoryToResolve)).then(({ children }) => {
+            if (children) {
+                const latestInput = this.pendingResolvedDirectories.get(directoryToResolve)?.input || '';
+                const childDirectories = children.filter(child => child.isDirectory)
+                    .map(directory => `${directory.resource.path}/`)
+                    .sort();
+                this.cachedDirectories.set(directoryToResolve, childDirectories);
+                this.directoryResolvedEmitter.fire({ parent: directoryToResolve, input: latestInput, children: childDirectories });
+            }
+        }).catch(e => {
+            // no-op
+        });
+    }
+}
 export class LocationListRenderer extends ReactRenderer {
 
+    protected directoryCache: ResolvedDirectoryCache;
+    protected toDisposeOnNewCache = new DisposableCollection();
     protected _drives: URI[] | undefined;
+    protected _doShowTextInput = false;
+    get doShowTextInput(): boolean {
+        return this._doShowTextInput;
+    }
+    set doShowTextInput(doShow: boolean) {
+        this._doShowTextInput = doShow;
+        if (doShow) {
+            this.initResolveDirectoryCache();
+        }
+    }
+    protected lastUniqueTextInputLocation: URI | undefined;
+    protected previousAutocompleteMatch: string;
+    protected doAttemptAutocomplete = true;
 
     constructor(
         protected readonly service: LocationService,
+        protected readonly fileService: FileService,
         host?: HTMLElement
     ) {
         super(host);
         this.doLoadDrives();
     }
 
     render(): void {
-        super.render();
+        ReactDOM.render(this.doRender(), this.host, this.doAfterRender);
+    }
+
+    protected initResolveDirectoryCache(): void {
+        this.toDisposeOnNewCache.dispose();
+        this.directoryCache = new ResolvedDirectoryCache(this.fileService);
+        this.toDisposeOnNewCache.push(this.directoryCache.onDirectoryDidResolve(({ parent, children, input }) => {
+            if (this.locationTextInput?.value === input) {
```

Comment:
This strategy of input matching seems a bit strange to me. Since `doRenderAutoSuggestion` works as long as the input is a child of the queried directory (and calculates the selection based on the current input, regardless of the `input` field from the `AutoSuggestDataEvent`), why not just check that here, and save the trouble of having to update (or even have)  the `input` field on `AutoSuggestDataEvent`? In fact, `doRenderAutoSuggestion` can mostly run regardless - it may just fail to find a matc, and end up doing nothing.

Reply:
Good catch, you're right about the redundant check in `doRenderAutoSuggestion`, which I've renamed to `tryRenderFirstMatch` for clarity. I've changed this a bit to verify that the parent value of the input field matches the parent of the event. This accounts for the scenario under high latency where you might be in the middle of a path and then press backspace several times, for example:

1. start typing:  `/home/kenma`     (auto suggest is pending children of `subdirectory/`)
1. press backspace several times: `/hom` (backspace does not trigger a *new* autosuggest attempt)
1.  autosuggest from step one resolves, but input field's direct parent is now `/` (suggestion will only render if input is a valid match)

Here is a high latency screen cast of the behavior:

![latency](https://user-images.githubusercontent.com/62660714/107567858-dd38d100-6bab-11eb-9b08-ece702b23387.gif)